### PR TITLE
feat: add exception code

### PR DESCRIPTION
**Pull request by aamingaa**
feat: add exception code

### CODE REVIEW

**Review by AlbumenJ — CHANGES_REQUESTED** [review]
Please fix uts

**Comment by aamingaa** [comment]
@AlbumenJ @EarthChen PTAL

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java [hunk 1]
```diff
@@ -46,14 +46,20 @@ public enum TripleHeaderEnum {
 
     TRI_HEADER_CONVERT("tri-header-convert"),
 
+    TRI_HEADER_EXCEPTION_CODE("tri-header-exception-code"),
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java:49]: `tri-exception-code` is better

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java [hunk 2]
```diff
@@ -46,14 +46,20 @@ public enum TripleHeaderEnum {
...
     ;
 
     static final Map<String, TripleHeaderEnum> enumMap = new HashMap<>();
 
     static final Set<String> excludeAttachmentsSet = new HashSet<>();
 
     static {
+        Set<String> needToDeliveryHeaders = needToDeliveryHeaders();
         for (TripleHeaderEnum item : TripleHeaderEnum.values()) {
+            if (needToDeliveryHeaders.contains(item.getHeader())) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java:60]: It's better not to make special judgments here.

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCall.java [hunk 3]
```diff
@@ -142,7 +163,11 @@ private void doSendMessage(Object message) {
         }
         final byte[] data;
         try {
-            data = packableMethod.packResponse(message);
+            if (message instanceof Exception && !isNeedReturnException) {
+                data = packableMethod.packResponse(new RpcException((Exception) message));
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCall.java:167]: Should exceptions be serialized directly?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java [hunk 4]
```diff
@@ -83,6 +83,9 @@ public void invoke() {
                     responseObserver.onCompleted(TriRpcStatus.DEADLINE_EXCEEDED);
                     return;
                 }
+                if (r.getValue() instanceof Throwable) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/AbstractServerCallListener.java:86]: Can the instanceof judgment be reduced?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/transport/AbstractH2TransportListener.java [hunk 5]
```diff
@@ -45,6 +45,26 @@ public abstract class AbstractH2TransportListener implements H2TransportListener
      * @param trailers the metadata from remote
      * @return KV pairs map
      */
+    protected Map<String, Object> exceptionHeadersToMap(Map<String, String> reserved, Supplier<Object> convertUpperHeaderSupplier) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/transport/AbstractH2TransportListener.java:48]: Can it be merged with headertoMap?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java [hunk 6]
```diff
@@ -55,12 +56,32 @@ public void onCancel(TriRpcStatus status) {
...
+                onReturnException(t);
+            } else {
+                onReturn(t);
+            }
         } else {
             super.doOnResponseHasException(t);
         }
     }
 
+    private void onReturnException(Object value) {
```
Comment (icodening) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java:69]: `Exception value` is better

**Review by icodening — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java [hunk 7]
```diff
@@ -55,12 +59,28 @@ public void onCancel(TriRpcStatus status) {
     @Override
     protected void doOnResponseHasException(Throwable t) {
         if (needWrapper) {
-            onReturn(t);
+            if (t instanceof Exception) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java:62]: Is judgment necessary here?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryClientCallListener.java [hunk 8]
```diff
@@ -42,7 +43,8 @@ public void onClose(TriRpcStatus status, Map<String, Object> trailers) {
         AppResponse result = new AppResponse();
         result.setObjectAttachments(trailers);
         if (status.isOk()) {
-            if (appResponse instanceof Exception) {
+           Integer exceptionCode = extractExceptionCode(trailers);
+            if (exceptionCode != 0) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryClientCallListener.java:47]: 0 should be a constant

**Review by EarthChen — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java [hunk 9]
```diff
@@ -70,6 +72,12 @@ public enum TripleHeaderEnum {
 
     }
 
+    public static Map<String, TripleHeaderEnum> needToDeliveryHeaders() {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleHeaderEnum.java:75]: unused ?

**Review by EarthChen — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java [hunk 10]
```diff
@@ -34,6 +35,9 @@ public UnaryServerCallListener(RpcInvocation invocation, Invoker<?> invoker,
 
     @Override
     public void onReturn(Object value) {
+        if (value instanceof Exception) {
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/UnaryServerCallListener.java:38]: Seems unnecessary?

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/ClientCall.java [hunk 11]
```diff
@@ -53,7 +53,7 @@ interface Listener {
          * @param status   response status
          * @param trailers response trailers
          */
-        void onClose(TriRpcStatus status, Map<String, Object> trailers);
+        void onClose(TriRpcStatus status, Map<String, Object> trailers, Map<String, String> triExceptionCodeAttachments);
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/call/ClientCall.java:56]: Is it possible to have more than one exceptionCode? Why use maps

**Review by EarthChen — CHANGES_REQUESTED** [review]

**Pull request closed without merging**